File: src/main/java/fr/spoutnik87/musicbot_rest/model/Category.java
package fr.spoutnik87.musicbot_rest.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonView;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Set;

@NoArgsConstructor
@RequiredArgsConstructor
@Data
@Entity
@Table(name = "Category")
public class Category extends AuditModel implements Serializable {

  @JsonView(Views.Public.class)
  @NonNull
  @JsonProperty("id")
  @Column(nullable = false, unique = true)
  private String uuid;

  @JsonView(Views.Public.class)
  @NonNull
  @Column(nullable = false)
  private String name;

  @JsonView(Views.Public.class)
  @OneToMany(mappedBy = "category", cascade = CascadeType.ALL)
  @JsonBackReference
  private Set<Media> mediaSet;

  @JsonView(Views.Public.class)
  @ManyToOne
  @JoinColumn(name = "server_id")
  @JsonManagedReference
  private Server server;
}

File: src/main/kotlin/fr/spoutnik87/musicbot_rest/model/Category.kt
package fr.spoutnik87.musicbot_rest.model

import java.io.Serializable
import javax.persistence.*

@Entity
@Table(name = "Category")
data class Category(
        @Column(nullable = false, unique = true)
        var uuid: String,
        @Column(nullable = false)
        var name: String,
        @Column(nullable = false)
        var thumbnailSize: Long
) : AuditModel(), Serializable {

    @ManyToOne
    @JoinColumn(name = "author_id")
    lateinit var author: User

    @ManyToOne
    @JoinColumn(name = "server_id")
    lateinit var server: Server

    constructor(uuid: String, name: String, thumbnailSize: Long, user: User, server: Server) : this(uuid, name, thumbnailSize) {
        this.author = user
        this.server = server
    }

    @OneToMany(mappedBy = "category", cascade = [CascadeType.ALL])
    val contentSet: MutableSet<Content> = HashSet()

    fun hasThumbnail() = thumbnailSize > 0
}